modules/aruco/src/apriltag_quad_thresh.cpp

```diff
@@ -0,0 +1,1606 @@
+            goto finish;
+    }
+
+    //    printf("%d %d %d %d\n", indices[0], indices[1], indices[2], indices[3]);
+
+    if (0) {
+        // no refitting here; just use those points as the vertices.
+        // Note, this is useful for debugging, but pretty bad in
+        // practice since this code path also omits several
+        // plausibility checks that save us tons of time in quad
+        // decoding.
+        for (int i = 0; i < 4; i++) {
+            struct pt *p;
+            _zarray_get_volatile(cluster, indices[i], &p);
+
+            quad->p[i][0] = .5*p->x; // undo fixed-point arith.
+            quad->p[i][1] = .5*p->y;
+        }
+
+        res = 1;
+
+    } else {
+        double lines[4][4];
+
+        for (int i = 0; i < 4; i++) {
+            int i0 = indices[i];
+            int i1 = indices[(i+1)&3];
+
+            if (0) {
+                // if there are enough points, skip the points near the corners
+                // (because those tend not to be very good.)
+                if (i1-i0 > 8) {
+                    int t = (i1-i0)/6;
+                    if (t < 0)
+                        t = -t;
+
+                    i0 = (i0 + t) % sz;
+                    i1 = (i1 + sz - t) % sz;
+                }
+            }
+
+            double err;
+            fit_line(lfps, sz, i0, i1, lines[i], NULL, &err);
+
+            if (err > _params->aprilTagMaxLineFitMse) {
+                res = 0;
+                goto finish;
+            }
+        }
+
+        for (int i = 0; i < 4; i++) {
+            // solve for the intersection of lines (i) and (i+1)&3.
+            // p0 + lambda0*u0 = p1 + lambda1*u1, where u0 and u1
+            // are the line directions.
+            //
+            // lambda0*u0 - lambda1*u1 = (p1 - p0)
+            //
+            // rearrange (solve for lambdas)
+            //
+            // [u0_x   -u1_x ] [lambda0] = [ p1_x - p0_x ]
+            // [u0_y   -u1_y ] [lambda1]   [ p1_y - p0_y ]
+            //
+            // remember that lines[i][0,1] = p, lines[i][2,3] = NORMAL vector.
+            // We want the unit vector, so we need the perpendiculars. Thus, below
+            // we have swapped the x and y components and flipped the y components.
+
+            double A00 =  lines[i][3],  A01 = -lines[(i+1)&3][3];
+            double A10 =  -lines[i][2],  A11 = lines[(i+1)&3][2];
+            double B0 = -lines[i][0] + lines[(i+1)&3][0];
+            double B1 = -lines[i][1] + lines[(i+1)&3][1];
+
+            double det = A00 * A11 - A10 * A01;
+
+            // inverse.
+            double W00 = A11 / det, W01 = -A01 / det;
+            if (fabs(det) < 0.001) {
+                res = 0;
+                goto finish;
+            }
+
+            // solve
+            double L0 = W00*B0 + W01*B1;
+
+            // compute intersection
+            quad->p[i][0] = lines[i][0] + L0*A00;
+            quad->p[i][1] = lines[i][1] + L0*A10;
+
+            if (0) {
```

Comment:
Enabling of this validation code breaks tests (assertion raised). Could you take a look on this?

Reply:
:thinking: you are right, i've compare it with the original code and it also breaks at that point...

Reply:
Looks like there is a typo below (A10 -> A01):
```.diff
-double x = lines[(i+1)&3][0] - L1*A10;
+double x = lines[(i+1)&3][0] - L1*A01;
```